src/main/java/com/rabbitmq/client/impl/ValueWriter.java

```diff
@@ -143,9 +143,17 @@ else if(value instanceof Integer) {
         else if(value instanceof BigDecimal) {
             writeOctet('D');
             BigDecimal decimal = (BigDecimal)value;
+            // The scale must be an unsigned octet, therefore its values must
+            // be between 0 and 255
+            if(decimal.scale() > 255 || decimal.scale() < 0)
+                throw new IllegalArgumentException
+                    ("BigDecimal has too large of a scale to be encoded. " +
+                            "The scale was: " + decimal.scale());
             writeOctet(decimal.scale());
             BigInteger unscaled = decimal.unscaledValue();
-            if(unscaled.bitLength() > 32) /*Integer.SIZE in Java 1.5*/
+            // We use 31 instead of 32 because bitLength ignores the sign bit,
+            // so e.g. new BigDecimal(Integer.MAX_VALUE) comes out to 31 bits.
+            if(unscaled.bitLength() > 31) /*Integer.SIZE in Java 1.5*/
```

Comment:
The comment about `Integer.SIZE` now no longer applies, IIUC.

Reply:
Good catch, fixed.